File: backend/services/cultivation_service.go
package services
import (
	"errors"
	"cronos-backend/blockchain"
	"cronos-backend/models"
	"cronos-backend/repositories"
	"github.com/google/uuid"
)
type CultivationService struct {
	CultivationRepo *repositories.CultivationRepository
	FarmRepo        *repositories.FarmRepository
	Ledger          *blockchain.Ledger
}
func NewCultivationService(cultRepo *repositories.CultivationRepository, farmRepo *repositories.FarmRepository, ledger *blockchain.Ledger) *CultivationService {
	return &CultivationService{CultivationRepo: cultRepo, FarmRepo: farmRepo, Ledger: ledger}
}
func (s *CultivationService) CreateCycle(userID uuid.UUID, cycle *models.CultivationCycle) error {
	pond, err := s.FarmRepo.FindPondByID(cycle.PondID)
	if err != nil {
		return errors.New("pond not found")
	}
	farm, err := s.FarmRepo.FindByID(pond.FarmID)
	if err != nil || farm.UserID != userID {
		return errors.New("unauthorized")
	}
	if err := s.CultivationRepo.Create(cycle); err != nil {
		return errors.New("failed to create cultivation cycle")
	}
	s.Ledger.RecordEvent("cultivation_started", userID, "cultivation", cycle.ID, map[string]interface{}{
		"pond_id":     cycle.PondID,
		"shrimp_type": cycle.ShrimpTypeID,
		"start_date":  cycle.StartDate,
	})
	return nil
}
func (s *CultivationService) GetMyCycles(userID uuid.UUID) ([]models.CultivationCycle, error) {
	return s.CultivationRepo.FindByUserID(userID)
}
func (s *CultivationService) GetCycle(id uuid.UUID) (*models.CultivationCycle, error) {
	return s.CultivationRepo.FindByID(id)
}
func (s *CultivationService) UpdateCycle(userID, cycleID uuid.UUID, req *models.CultivationCycle) (*models.CultivationCycle, error) {
	cycle, err := s.CultivationRepo.FindByID(cycleID)
	if err != nil {
		return nil, errors.New("cycle not found")
	}
	if req.Status != "" {
		cycle.Status = req.Status
	}
	if req.ActualEndDate != nil {
		cycle.ActualEndDate = req.ActualEndDate
	}
	if req.Notes != "" {
		cycle.Notes = req.Notes
	}
	if err := s.CultivationRepo.Update(cycle); err != nil {
		return nil, errors.New("failed to update cycle")
	}
	return cycle, nil
}
func (s *CultivationService) AddFeedLog(userID uuid.UUID, log *models.FeedLog) error {
	if err := s.CultivationRepo.CreateFeedLog(log); err != nil {
		return errors.New("failed to add feed log")
	}
	s.Ledger.RecordEvent("feed_logged", userID, "cultivation", log.CultivationCycleID, map[string]interface{}{
		"feed_type": log.FeedType,
		"quantity":  log.Quantity,
	})
	return nil
}
func (s *CultivationService) GetFeedLogs(cycleID uuid.UUID) ([]models.FeedLog, error) {
	return s.CultivationRepo.FindFeedLogs(cycleID)
}
func (s *CultivationService) AddWaterQualityLog(userID uuid.UUID, log *models.WaterQualityLog) error {
	if err := s.CultivationRepo.CreateWaterQualityLog(log); err != nil {
		return errors.New("failed to add water quality log")
	}
	s.Ledger.RecordEvent("water_quality_logged", userID, "cultivation", log.CultivationCycleID, map[string]interface{}{
		"temperature":      log.Temperature,
		"ph":               log.PH,
		"salinity":         log.Salinity,
		"dissolved_oxygen": log.DissolvedOxygen,
	})
	return nil
}
func (s *CultivationService) GetWaterQualityLogs(cycleID uuid.UUID) ([]models.WaterQualityLog, error) {
	return s.CultivationRepo.FindWaterQualityLogs(cycleID)
}

File: backend-spring/src/main/java/com/cronos/service/CultivationService.java
package com.cronos.service;

import com.cronos.blockchain.BlockchainLedger;
import com.cronos.entity.*;
import com.cronos.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import java.util.*;

@Service
@RequiredArgsConstructor
public class CultivationService {
    private final CultivationCycleRepository cultivationRepo;
    private final FeedLogRepository feedLogRepo;
    private final WaterQualityLogRepository waterQualityRepo;
    private final FarmRepository farmRepo;
    private final PondRepository pondRepo;
    private final BlockchainLedger ledger;

    public CultivationCycle createCycle(UUID userId, CultivationCycle cycle) {
        Pond pond = pondRepo.findById(cycle.getPondId()).orElseThrow(() -> new RuntimeException("pond not found"));
        Farm farm = farmRepo.findById(pond.getFarmId()).orElseThrow(() -> new RuntimeException("farm not found"));
        if (!farm.getUserId().equals(userId)) throw new RuntimeException("unauthorized");
        cultivationRepo.save(cycle);
        ledger.recordEvent("cultivation_started", userId, "cultivation", cycle.getId(),
                Map.of("pond_id", cycle.getPondId(), "shrimp_type", cycle.getShrimpTypeId(), "start_date", cycle.getStartDate().toString()));
        return cycle;
    }

    public List<CultivationCycle> getMyCycles(UUID userId) { return cultivationRepo.findByUserId(userId); }
    public CultivationCycle getCycle(UUID id) { return cultivationRepo.findById(id).orElseThrow(() -> new RuntimeException("cycle not found")); }

    public CultivationCycle updateCycle(UUID userId, UUID cycleId, CultivationCycle req) {
        CultivationCycle cycle = cultivationRepo.findById(cycleId).orElseThrow(() -> new RuntimeException("cycle not found"));
        if (req.getStatus() != null && !req.getStatus().isEmpty()) cycle.setStatus(req.getStatus());
        if (req.getActualEndDate() != null) cycle.setActualEndDate(req.getActualEndDate());
        if (req.getNotes() != null && !req.getNotes().isEmpty()) cycle.setNotes(req.getNotes());
        cultivationRepo.save(cycle);
        return cycle;
    }

    public FeedLog addFeedLog(UUID userId, FeedLog log) {
        feedLogRepo.save(log);
        ledger.recordEvent("feed_logged", userId, "cultivation", log.getCultivationCycleId(),
                Map.of("feed_type", log.getFeedType(), "quantity", log.getQuantity()));
        return log;
    }

    public List<FeedLog> getFeedLogs(UUID cycleId) { return feedLogRepo.findByCultivationCycleIdOrderByFeedingTimeDesc(cycleId); }

    public WaterQualityLog addWaterQualityLog(UUID userId, WaterQualityLog log) {
        waterQualityRepo.save(log);
        ledger.recordEvent("water_quality_logged", userId, "cultivation", log.getCultivationCycleId(),
                Map.of("temperature", log.getTemperature(), "ph", log.getPh(), "salinity", log.getSalinity(), "dissolved_oxygen", log.getDissolvedOxygen()));
        return log;
    }

    public List<WaterQualityLog> getWaterQualityLogs(UUID cycleId) { return waterQualityRepo.findByCultivationCycleIdOrderByRecordedAtDesc(cycleId); }
}
